clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -93,81 +135,114 @@ func (a *apiAvailabilityMeasurement) start(config *measurement.Config, SSHToMast
 			select {
 			case <-a.stopCh:
 				return
-			case <-time.After(time.Duration(probeDuration)):
+			case <-a.holdCh:
+				select {
+				case <-a.stopCh:
+					return
+				case <-a.holdCh:
+					continue
+				}
+			case <-time.After(a.frequency):
 				a.updateClusterAvailabilityMetrics(k8sClient)
-				if SSHToMasterSupported {
-					a.updateMasterAvailabilityMetrics(k8sClient, config, provider)
+				if a.masterLevelAvailabilityEnabled() {
+					a.updateMasterAvailabilityMetrics(k8sClient, provider)
 				}
 			}
 		}
 	}()
 	return nil
 }
 
-// Execute starts the api-server healthz probe end point from start action and
-// collects availability metrics in gather.
-func (a *apiAvailabilityMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
-	SSHToMasterSupported := !config.ClusterFramework.GetClusterConfig().Provider.Features().SupportSSHToMaster
-
-	if !SSHToMasterSupported {
-		klog.Infof("ssh to master nodes not supported. Measurement would have only Cluster Level Metrics")
+func (a *apiAvailabilityMeasurement) initFields(config *measurement.Config) error {
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.holdCh = make(chan struct{})
+	frequency, err := util.GetDuration(config.Params, "frequency")
+	if err != nil {
+		return err
+	}
+	a.frequency = frequency
+
+	a.clusterLevelMetrics = &apiAvailabilityMetrics{}
+	if config.ClusterLoaderConfig.EnableExecService {
+		a.masterIPs = config.ClusterFramework.GetClusterConfig().MasterInternalIPs
+		if len(a.masterIPs) == 0 {
+			klog.V(2).Infof("%s: master internal IP(s) are not provided, therefore only cluster-level availability will be measured", a)
+			return nil
+		}
+		a.masterLevelMetrics = map[string]*apiAvailabilityMetrics{}
+		for _, ip := range a.masterIPs {
+			a.masterLevelMetrics[ip] = &apiAvailabilityMetrics{}
+		}
+	} else {
+		klog.V(2).Infof("%s: exec service is not enabled, therefore only cluster-level availability will be measured", a)
 	}
 
-	action, err := util.GetString(config.Params, "action")
-	if err != nil {
-		return nil, err
+	return nil
+}
+
+func (a *apiAvailabilityMeasurement) masterLevelAvailabilityEnabled() bool {
+	return len(a.masterLevelMetrics) > 0
+}
+
+func (a *apiAvailabilityMeasurement) hold() {
+	if !a.isRunning {
+		klog.V(2).Infof("%s: measurement is not running", a)
+		return
 	}
-	probeFrequency, err := util.GetIntOrDefault(config.Params, "frequency", 1)
-	if err != nil {
-		return nil, err
+	if a.isHeld {
```

Comment:
There is a potential race here. If I have a measurement step with two measurements that hold (measurements are executed in parallel) I may either
1. Hold the measurement and get "already held" error <- OK
2. Hold and "unhold" the measurement. The a.isHeld will be true, but measurement will continue running <- not OK, program will be in a malformed state.

This is because you allow two or more goroutines to modify the a.isHeld. You could fix that with either locks or (more golang idiomatic) having the a.IsHeld being modified only inside a single go-routine. You're already have this go-routine in the start() method. So the code could be something like:

```
func (a *apiAvailabilityMeasurement) hold() {
  a.hold <- struct{}{}
}

func (a *apiAvailabilityMeasurement) unhold() {
  a.unhold <- struct{}{}
}

func (a *apiAvailabilityMeasurement) start() {
  ...
  go func() {
    for {
      if (a.isHeld) {
        select {
        case <- a.unholdCh:
          a.isHeld = false
        case <- a.holdCh:
           klog.Warning("Meaasurement is already held!")
        case <- a.stopCh:
          return
        }

         
      }
      select {
        ...
        case <- a.holdCh:
          a.isHeld = true
          continue
        case <- a.unHoldCh:
          klog.Warning("Meaasurement is already unheld!")
      }
    }
  }
}
```

I think it's safer and also easier to read.

Reply:
I used the lock to prevent two actions of the same measurement to be run in the same step, so we don't run into any race by accidental putting two actions of the same measurement into one step.

I really like the idea of separating the pausing/unpausing channel into two different ones, though, as it indeed increases the readability of the code. Thanks for that remark.